src/pip/_internal/utils/setuptools_build.py

```diff
@@ -12,8 +12,8 @@
 # invoking via the shim.  This avoids e.g. the following manifest_maker
 # warning: "warning: manifest_maker: standard file '-c' not found".
 _SETUPTOOLS_SHIM = (
-    "import sys, setuptools, tokenize; sys.argv[0] = {0!r}; __file__={0!r};"
-    "f=getattr(tokenize, 'open', open)(__file__);"
+    "import sys, setuptools.build_meta, tokenize; sys.argv[0] = {0!r}; __file__={0!r};"
+    "f = setuptools.build_meta._open_setup_script(__file__) if hasattr(setuptools.build_meta, '_open_setup_script') else getattr(tokenize, 'open', open)(__file__);"
```

Comment:
Instead of trying to reply on setuptools internals, I would just do something like

```python
import io, os, sys, setuptools, tokenize
f = getattr(tokenize, 'open', open)(__file__) if os.path.exists(__file__) else io.StringIO(u'from setuptools import setup; setup()')
```

Reply:
Oops, race condition with @pfmoore!

Reply:
I just realised this can be simplified further since we’ve dropped 2.7 support:

```python
import io, os, sys, setuptools, tokenize
f = tokenize.open(__file__) if os.path.exists(__file__) else io.StringIO(u'from setuptools import setup; setup()')
```

Reply:
Fixed.